### DISCUSSION TITLE
assign not working as expected

### DISCUSSION THREAD

**Discussion by totalard** [mikro-orm/mikro-orm#3058] [Q&A]
**Describe the bug**
using wrap(e).assing repo.assign em.assign none of them working as expected.

all string, arrays, bools,ints...etc. are assigning to managed entity and persistAndFlush save to db, 

but object field like 1:1 M:1 not assigning, i sending whole object as body, when worked in typeorm, it compare and assign them to entity and save will trigger save to db, here assign is the only issue i found, 

creating new entity work fine, i have issue only in update

**Expected behavior**
assign will assign all post data to entity respect to keys, and allow us to save, if ID found, it update it, if ID is undefined, it should create it

**Additional context**
```
async upsert(entity: any) {
    let e: T;
    if (entity.id === undefined) {
      e = this.genericRepository.create(entity);
    } else {
      e = await this.genericRepository.findOne({
        id: entity.id,
      } as FilterQuery<T>);
      this.genericRepository.assign(e, entity);
    }
    await this.genericRepository.persistAndFlush(e);
    return e;
  }
```
i also tried wrap(e).assign(entity,{nested...true, mergeobject..true, etc..})  no success

i am using all latest version, mongodb, just installed today

**Comment by B4nan** [comment]
Please provide more details, ideally something executable, as I have no idea what data are you actually passing, also no entity definition. Note that you should not use entity instance in the assign parameters.

**Reply by totalard** [reply]
in wrap(e).assign(entity)

the e is entity instance
entity is just body passing from formData

https://jsonformatter.org/json-pretty-print/7d4a07 

check the formdata i submitting

the formdata include whole object, as assign only use required params, like other orms do.

**Please Note, I can update all string, bool, number, even relation arrays,, but not object**

check the property images, i can create, But cant update, cant change image property at all


```
import { Review } from '../../reviews/entities/review.entity';
import { Tag } from '../../tag/entities/tag.entity';
import { Category } from '../../category/entities/category.entity';
import { Media } from 'src/modules/file-manager/entities/media.entity';
import { BaseWithSlug } from 'src/modules/baseWithSlug.entity';
import { ProductAttribute } from './productAttribute.entity';
import { ProductVariation } from './ProductVariation.entity';
import {
  Collection,
  Entity,
  ManyToMany,
  ManyToOne,
  OneToMany,
  Property,
  t,
} from '@mikro-orm/core';

@Entity()
export class Product extends BaseWithSlug {
  @Property({ nullable: true })
  discription: string;

  @Property({ nullable: true })
  short: string;

  @Property({ nullable: true })
  publicNote: string;

  @Property()
  sku: string;

  @Property({ type: t.float, default: 0.0 })
  regPrice: number;

  @Property({ type: t.float, default: 0.0 })
  salePrice: number;

  @Property({ type: t.float, default: 0.0 })
  stock: number;

  @Property()
  status: string = 'draft';

  @Property()
  isFeatured: boolean;

  @Property()
  meta: object = {};

  @OneToMany(() => Review, (review) => review.product, { orphanRemoval: true })
  reviews = new Collection<Review>(this);

  @Property({ type: t.float, default: 0.0 })
  weight: number;

  @Property({ type: t.float, default: 0.0 })
  length: number;

  @Property({ type: t.float, default: 0.0 })
  hight: number;

  @Property({ type: t.float, default: 0.0 })
  width: number;

  @Property()
  type: string;

  @ManyToMany({ entity: () => Tag, mappedBy: 'products' })
  tags = new Collection<Tag>(this);

  @OneToMany(() => ProductAttribute, (attr) => attr.product, {
    orphanRemoval: true,
  })
  attributes = new Collection<ProductAttribute>(this);

  @ManyToMany(() => Category)
  categories = new Collection<Category>(this);

  @ManyToMany({ entity: () => Media, mappedBy: 'products' })
  images = new Collection<Media>(this);

  @Property({ nullable: true })
  deletedAt?: Date;

  @ManyToOne(() => Media, { nullable: true })
  image: Media;

  @ManyToMany(() => Product)
  upSell = new Collection<Product>(this);

  @ManyToMany(() => Product)
  crossSell = new Collection<Product>(this);

  @OneToMany(() => ProductVariation, (variation) => variation.product, {
    orphanRemoval: true,
  })
  varProducts = new Collection<ProductVariation>(this);
}
```

The Way i Implemented Upsert, in Which create Works fine, in Update as i already mentioned, except Object , all works fine, 

```
async upsert(entity: any) {
    let e: T;
    if (entity.id === undefined) {
      e = this.genericRepository.create(entity);
    } else {
      e = await this.genericRepository.findOne({
        id: entity.id,
      } as FilterQuery<T>);
      this.genericRepository.assign(e, entity);
    }
    await this.genericRepository.persistAndFlush(e);
    return e;
  }
```

Also Note. this works fine, but it is **generic**, i cant explicitly cant do such way, as the upsert is generic, why only object M:1 (confirmed), 1:1(not checked) not assigning in update, any clue, putting populate:true make nodejs crash due to heavy populating, so i cant do populate:true, but i explicitely added "image" in populate to test, still no success

```
async upsert(entity: any) {
    let e: T;
    if (entity.id === undefined) {
      e = this.genericRepository.create(entity);
    } else {
      e = await this.genericRepository.findOne({
        id: entity.id,
      } as FilterQuery<T>);
      this.genericRepository.assign(e, entity);
     e.image = entity.image;
    }
    await this.genericRepository.persistAndFlush(e);
    return e;
  }
```

**Reply by totalard** [reply]
https://github.com/mikro-orm/mikro-orm/discussions/2920#discussioncomment-2371339

Why I can't provide existing entity ?? Only FK as string is accepted . Can u update on this ??

**Reply by totalard** [reply]
The Problem Temporary Solved By this workaround,

Here e is Entity and entity is POST body / formdata
```
async upsert(entity: any) {
    let e: T;
    if (entity.id === undefined) {
      e = this.genericRepository.create(entity);
    } else {
      e = await this.genericRepository.findOne({
        id: entity.id,
      } as FilterQuery<T>);
      Object.keys(entity).forEach((each) => {
        if (
          typeof entity[each] === 'object' &&
          entity[each]?.id !== undefined
        ) {
          entity[each] = entity[each]?.id;
        }
      });
      this.genericRepository.assign(e, entity);
    }
    await this.genericRepository.flush();
    return e;
  }
```

in M:N, !:M , its saved correctly using full entitydata, but in object realtions M:1, 1:1 its not like that, is it a bug ? assign should act uniformly on all relations, you must fix it, so it become very easy to post

**Reply by B4nan** [reply]
I'm having hard times understanding your english. And saying "you must fix this" seems quite rude, I am not obliged to fix your problems.

What you are probably missing is that you _need_ to have entities already loaded in context, otherwise you can't assign to them. This applies to all relations you want to assign on. `assign` is a sync method, it can never load anything. And to update any entity, it first needs to be loaded. 

Also note that there was one bug in `assign` that is fixed in master, so you might wanna try the dev version: https://github.com/mikro-orm/mikro-orm/commit/90ec83fc01296dc6f3560df00e637b050852b7f2. 

If you want to speed things up, try to prepare a complete but minimal reproduction, so far you provided only bits and pieces that are far away from being minimal or complete.

**Reply by totalard** [reply]
very before, around 1.5 year, i left mikro due to some assign problem, that i cant fix and really cant understand. i already asked many questions regarding this,  mikro in 2 or 3 that time,

now i returned to mikro, still facing such problem, so i said you must fix it,

i found a work around, and its working now, just some adjustment,
